lib/cli/ui/prompt/interactive_options.rb

```diff
@@ -68,61 +79,96 @@ def call
-            num_lines.times { print(ANSI.previous_line) }
+            number_of_lines.times { print(ANSI.previous_line) }
           end
         end
 
-        def clear_output
+        def clear_output(number_of_lines=num_lines)
           CLI::UI.raw do
             # Write over all lines with whitespace
-            num_lines.times { puts(' ' * CLI::UI::Terminal.width) }
+            number_of_lines.times { puts(' ' * CLI::UI::Terminal.width) }
           end
-          reset_position
-        end
+          reset_position number_of_lines
 
-        def num_lines
-          options = presented_options.map(&:first)
+          # Update if metadata is being displayed
+          # This must be done _after_ the output is cleared or it won't draw over
+          # the entire output
+          @displaying_metadata = display_metadata?
+        end
 
-          # empty_option_count is needed since empty option titles are omitted
-          # from the line count when reject(&:empty?) is called
-          empty_option_count = options.count(&:empty?)
+        # Don't use this in place of +@displaying_metadata+, this updates too
+        # quickly to be useful when drawing to the screen.
+        def display_metadata?
+          filtering? or selecting? or has_filter?
+        end
 
-          # options will be an array of questions but each option can be multi-line
-          # so to get the # of lines, you need to join then split
+        def num_lines
+          calculate_option_line_lengths if terminal_width_changed?
 
-          # since lines may be longer than the terminal is wide, we need to
-          # determine how many extra lines would be taken up by them
-          max_width = (CLI::UI::Terminal.width -
-                       options.count.to_s.size - # Width of the displayed number
-                       5 -                       # Extra characters added during rendering
-                       (@multiple ? 1 : 0)       # Space for the checkbox, if rendered
-                      ).to_f
+          option_length = presented_options.reduce(0) do |total_length, (_, option_number)|
+            # Handle continuation markers and "Done" option when multiple is true
+            next total_length + 1 if option_number.nil? or option_number.zero?
+            total_length + @option_lengths[option_number - 1]
+          end
 
-          total_non_empty_lines = options
-                                    .join("\n")
-                                    .split("\n")
-                                    .reject(&:empty?)
-                                    .map { |l| (l.length / max_width).ceil }
-                                    .reduce(&:+)
+          option_length + (@displaying_metadata ? 1 : 0)
+        end
 
-          total_non_empty_lines + empty_option_count
+        def terminal_width_changed?
+          @terminal_width_at_calculation_time != CLI::UI::Terminal.width
         end
 
         ESC = "\e"
+        BACKSPACE = "\u007F"
```

Comment:
It initially strikes me as slightly odd that we'd be dealing with UTF-8 here rather than arbitrary bytes. Anyone have an opinion on this?